src/Store/ExtensionClientStoreConnector.re

```diff
@@ -455,6 +463,54 @@ let start = (extensions, setup: Setup.t) => {
       )
     });
 
+  let getOriginalUri = (bufferId, path, providers: list(SCM.Provider.t)) =>
+    Isolinear.Effect.createWithDispatch(~name="scm.getOriginalUri", dispatch => {
+      // Try our luck with every provider. If several returns Last-Writer-Wins
+      // TODO: Is there a better heuristic? Perhaps use rootUri to choose the "nearest" provider?
+      providers
+      |> List.iter((provider: SCM.Provider.t) => {
+           let promise =
+             ExtHostClient.provideOriginalResource(
+               provider.handle,
+               Uri.fromPath(path),
+               extHostClient,
+             );
+
+           Lwt.on_success(promise, uri =>
+             dispatch(Actions.SCM(SCM.GotOriginalUri({bufferId, uri})))
+           );
+         })
+    });
+
+  let getOriginalContent = (bufferId, uri, providers) =>
+    Isolinear.Effect.createWithDispatch(
+      ~name="scm.getOriginalSourceLines", dispatch => {
+      let scheme = uri |> Uri.getScheme |> Uri.Scheme.toString;
+      providers
+      |> List.find_opt(((_, providerScheme)) => providerScheme == scheme)
+      |> Option.iter(provider => {
+           let (handle, _) = provider;
+           let promise =
+             ExtHostClient.provideTextDocumentContent(
+               handle,
+               uri,
+               extHostClient,
+             );
+
+           Lwt.on_success(
+             promise,
+             content => {
+               let lines =
+                 content |> Str.(split(regexp_string("\n"))) |> Array.of_list;
```

Comment:
Do we need to handle Windows-style line ending (CRLF) here?

Reply:
Isn't that covered by `\n`? Windows is just `\r\n` right? How do you handle it elsewhere?

Reply:
Right, Windows could be`\r\n`. The concern I had would be that, our `Buffer.t` / `BufferLine.t` don't encode any newlines. 

So if we trim on the `\n`, there might still be a trailing `\r` that would show up as a diff (an extra character) - since the buffer/bufferlines we use don't store any newlines.

Reply:
Forgot about this. Fixed in https://github.com/onivim/oni2/pull/1262/commits/58c237fe70d692b356ee6171b23af8eb52276501